iree/compiler/Conversion/LinalgToSPIRV/VectorizeMemref.cpp

```diff
@@ -0,0 +1,307 @@
+//===- VectorizeMemref.cpp - Convert memref of scalar to vector------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// Pass to convert memref into memref of vector.
+//
+//===----------------------------------------------------------------------===//
+#include "iree/compiler/Conversion/LinalgToSPIRV/Passes.h"
+#include "mlir/Dialect/StandardOps/IR/Ops.h"
+#include "mlir/Dialect/Vector/VectorOps.h"
+
+using namespace mlir;
+
+constexpr int vectorizationSizeInBits = 128;
+
+/// Return true of the type is a memref that can be vectorized to vector<4xi32>
+static bool memRefCanBeVectorized(Type type) {
+  auto memrefType = type.dyn_cast<MemRefType>();
+  if (!memrefType) return false;
+  if (memrefType.getElementType().isa<VectorType>()) return false;
+  if (vectorizationSizeInBits % memrefType.getElementTypeBitWidth())
+    return false;
+  if (ShapedType::isDynamic(memrefType.getShape().back())) return false;
+  if ((memrefType.getElementTypeBitWidth() * memrefType.getShape().back()) %
+          vectorizationSizeInBits !=
+      0)
+    return false;
+  return true;
+}
+
+/// Return the bitwith of a scalar or vector type.
+static unsigned getBitWidth(Type type) {
+  assert((type.isIntOrFloat() || type.isa<VectorType>()) &&
+         "bitwidth is not supported for this type");
+  if (type.isIntOrFloat()) return type.getIntOrFloatBitWidth();
+  auto vecType = type.cast<VectorType>();
+  auto elementType = vecType.getElementType();
+  return elementType.getIntOrFloatBitWidth() * vecType.getNumElements();
+}
+
+namespace {
+/// Analysis memref usages to decide if it should be vectorized. Right now the
+/// logic is to vectorize memref only if it is used by
+/// vectortransfer_read/vectortransfer_write operations.
+class MemRefUsageAnalysis {
+ public:
+  explicit MemRefUsageAnalysis(mlir::Operation *);
+
+  // Return true if the operation should be lowered using operations on
+  // cooperative matrix type.
+  bool vectorizeMemRef(Value v) const { return vectorize.count(v); }
```

Comment:
Is this for operations related to co-operative matrix or is it just using vector types. Maybe the method name can be modified to be clearer?